File: src/minima_harness/minima/memory.py
"""HarnessMemory — the seam that lets a :class:`MinimaAgent` *use* Mubit memory, not just
Minima's model recommender.

The harness already *reads* Mubit (lessons + optimized prompt are injected via
``tui.mubit.effective_prompt``) but never *writes*: ``init_mubit`` sets ``auto_learn=False``
and nothing calls ``remember``/``outcome``/``reflect``. So the loop is open — lessons are
consumed but the agent's own coding outcomes never feed back, and the injected lessons
never improve. This seam closes it, mirroring how the server-side memory adapter does:

- :meth:`recall` — task-specific prior context, injected before the model runs.
- :meth:`record_outcome` — a trace + an outcome score, *attributed to the recommendation*
  (``idempotency_key``/``reference_id`` = ``recommendation_id``). Never fabricates: a judge
  abstention (``quality is None``) records the trace but no score.
- :meth:`end_session` — ``reflect`` + ``checkpoint`` so a run leaves distilled, durable memory.

The default is :class:`NoopHarnessMemory`, so :class:`MinimaAgent` behaves exactly as before
unless a :class:`MubitHarnessMemory` is wired in. The Mubit-backed impl is fail-open (memory
must never break a coding run) and bridges the *synchronous* mubit SDK off the event loop.
"""

from __future__ import annotations

import logging
from typing import Protocol, runtime_checkable

import anyio

_log = logging.getLogger("minima_harness.memory")

# Entry types worth recalling before a coding turn (distilled signal, not raw traces).
_RECALL_ENTRY_TYPES = ["lesson", "rule", "observation"]
# Hard caps so injected context can never blow the model's window or run away on cost.
_MAX_SNIPPETS = 5
_MAX_SNIPPET_CHARS = 240


@runtime_checkable
class HarnessMemory(Protocol):
    """The memory a :class:`MinimaAgent` talks to. Implementations must be fail-open."""

    async def recall(self, task: str, *, limit: int = _MAX_SNIPPETS) -> list[str]:
        """Short prior-context snippets relevant to ``task`` (may be empty)."""
        ...

    async def record_outcome(
        self,
        *,
        task: str,
        recommendation_id: str,
        model_id: str,
        outcome: str,
        quality: float | None,
        cost_usd: float,
        latency_ms: int,
        turns: int,
    ) -> None:
        """Persist this turn's realized outcome, attributed to ``recommendation_id``."""
        ...

    async def end_session(self) -> None:
        """Distil the session into durable memory (reflect + checkpoint)."""
        ...


class NoopHarnessMemory:
    """Memory disabled — the default. Every method is a no-op / empty."""

    async def recall(self, task: str, *, limit: int = _MAX_SNIPPETS) -> list[str]:
        return []

    async def record_outcome(
        self,
        *,
        task: str,
        recommendation_id: str,
        model_id: str,
        outcome: str,
        quality: float | None,
        cost_usd: float,
        latency_ms: int,
        turns: int,
    ) -> None:
        return None

    async def end_session(self) -> None:
        return None


def format_recall_block(snippets: list[str]) -> str:
    """Render recalled snippets as a compact, clearly-delimited system-prompt section."""
    lines = "\n".join(f"- {s}" for s in snippets)
    return (
        '<prior_learnings source="mubit">\n'
        "Lessons and outcomes from earlier work on this project; apply when relevant.\n"
        f"{lines}\n"
        "</prior_learnings>"
    )


def _snippet_of(entry: object) -> str:
    """Best-effort text of a recall entry (object attrs or dict), capped."""
    for attr in ("content", "text", "summary"):
        val = getattr(entry, attr, None)
        if isinstance(val, str) and val.strip():
            return val.strip()[:_MAX_SNIPPET_CHARS]
    if isinstance(entry, dict):
        for key in ("content", "text", "summary"):
            dval = entry.get(key)
            if isinstance(dval, str) and dval.strip():
                return dval.strip()[:_MAX_SNIPPET_CHARS]
    if isinstance(entry, str) and entry.strip():
        return entry.strip()[:_MAX_SNIPPET_CHARS]
    return ""


class MubitHarnessMemory:
    """Mubit-SDK-backed memory: session-scoped, fail-open, sync-SDK bridged off the loop.

    Relies on the module-level mubit context set up by ``tui.mubit.init_mubit``. Availability
    is checked per call so wiring order (init runs *after* the agent is built) is irrelevant —
    before init, every method degrades to the no-op path.
    """

    def __init__(self, *, session_id: str, agent_id: str = "minima-harness") -> None:
        self._session_id = session_id
        self._agent_id = agent_id

    def _ready(self) -> bool:
        try:
            from minima_harness.tui.mubit import available

            return available()
        except Exception:  # noqa: BLE001 - never let a probe break the caller
            return False

    async def recall(self, task: str, *, limit: int = _MAX_SNIPPETS) -> list[str]:
        if not task.strip() or not self._ready():
            return []
        cap = max(1, min(limit, _MAX_SNIPPETS))
        try:
            import mubit

            # No session_id: recall across the whole project (agent_id + project_id set at
            # init_mubit), so THIS run learns from PAST sessions on the repo — the point of a
            # coding agent's memory. Writes/reflect below stay session-scoped.
            entries = await anyio.to_thread.run_sync(
                lambda: mubit.recall(task, limit=cap, entry_types=_RECALL_ENTRY_TYPES)
            )
        except Exception:  # noqa: BLE001 - recall must never break a run
            _log.debug("mubit_recall_failed", exc_info=True)
            return []
        snippets: list[str] = []
        for entry in entries or []:
            text = _snippet_of(entry)
            if text:
                snippets.append(text)
            if len(snippets) >= cap:
                break
        return snippets

    async def record_outcome(
        self,
        *,
        task: str,
        recommendation_id: str,
        model_id: str,
        outcome: str,
        quality: float | None,
        cost_usd: float,
        latency_ms: int,
        turns: int,
    ) -> None:
        if not recommendation_id or not self._ready():
            return
        qtext = "n/a" if quality is None else round(quality, 3)
        trace = (
            f"model={model_id} outcome={outcome} quality={qtext} "
            f"cost_usd={round(cost_usd, 6)} latency_ms={latency_ms} turns={turns} "
            f":: {task.strip()[:160]}"
        )

        def _write() -> None:
            import mubit

            # The trace, attributed to the recommendation for provenance + dedup.
            mubit.remember(
                trace,
                intent="trace",
                session_id=self._session_id,
                agent_id=self._agent_id,
                idempotency_key=recommendation_id,
            )
            # Close the loop with a *real* outcome score tied to the recommendation. Never
            # fabricate — a judge abstention (quality is None) records the trace but no score,
            # so Mubit's reflection learns only from graded turns.
            if quality is not None:
                mubit.outcome(
                    float(quality),
                    outcome_label=outcome,
                    reference_id=recommendation_id,
                    session_id=self._session_id,
                )

        try:
            await anyio.to_thread.run_sync(_write)
        except Exception:  # noqa: BLE001 - write-back must never break a run
            _log.debug("mubit_record_outcome_failed", exc_info=True)

    async def end_session(self) -> None:
        if not self._ready():
            return

        def _close() -> None:
            import mubit

            # reflect() distils this session's traces/outcomes into lessons; checkpoint()
            # leaves a durable marker so a future run can reference where this one ended.
            mubit.reflect(session_id=self._session_id)
            mubit.checkpoint(label="session-end", session_id=self._session_id)

        try:
            await anyio.to_thread.run_sync(_close)
        except Exception:  # noqa: BLE001 - session teardown must never break shutdown
            _log.debug("mubit_end_session_failed", exc_info=True)

File: packages/tui/src/minima/memory.ts
/**
 * HarnessMemory — lets a MinimaAgent *use* Mubit memory, not just Minima's model
 * recommender. Port of the Python harness's minima/memory.py to the TS harness (the binary
 * shipped users actually run).
 *
 * The harness read Mubit lessons but never wrote outcomes back, so the loop was open.
 * This closes it:
 *  - recall(task)      -> task-relevant prior context, injected before the model runs.
 *  - recordOutcome(..) -> a trace + an outcome score, attributed to the recommendation
 *    (idempotency_key / reference_id = recommendationId). Never fabricates: a judge
 *    abstention (quality === null) records the trace but no score.
 *  - endSession()      -> reflect + checkpoint, distilling a run into durable memory.
 *
 * Default is NoopHarnessMemory, so MinimaAgent behaves exactly as before unless a
 * MubitHarnessMemory is wired in. The Mubit-backed impl is fail-open (memory must never
 * break a coding run) and talks to the @mubit-ai/sdk Client through a small structural
 * interface, so this module + its tests need neither the SDK nor a network.
 */

const RECALL_ENTRY_TYPES = ["lesson", "rule", "observation"];
const MAX_SNIPPETS = 5;
const MAX_SNIPPET_CHARS = 240;

/**
 * Write-side lane partition. Two structurally incompatible writers share one Mubit
 * backend: the Minima SERVER writes typed OutcomeRecord JSON (intent=observation,
 * lane=minima:<namespace>, its own run ids) and this HARNESS writes free-text traces.
 * Stamping every harness write with its own lane keeps the two deliberately
 * partitioned at the source, whatever run ids are in play. Recall is deliberately
 * NOT lane-filtered yet — pre-partition entries have no lane and would vanish from
 * recall; filter once unlaned data has aged out.
 */
const HARNESS_LANE = "harness";

export interface OutcomeRecord {
  task: string;
  recommendationId: string;
  modelId: string;
  outcome: string;
  quality: number | null;
  costUsd: number;
  latencyMs: number;
  turns: number;
}

export interface HarnessMemory {
  /** Short prior-context snippets relevant to `task` (possibly empty). */
  recall(task: string, limit?: number): Promise<string[]>;
  /** Persist this turn's realized outcome, attributed to its recommendation. */
  recordOutcome(record: OutcomeRecord): Promise<void>;
  /** Distil the session into durable memory (reflect + checkpoint). */
  endSession(): Promise<void>;
}

/** The subset of the @mubit-ai/sdk Client this module uses (structural — for injection). */
export interface MemoryClient {
  recall(req: Record<string, unknown>): Promise<unknown>;
  remember(req: Record<string, unknown>): Promise<unknown>;
  recordOutcome(req: Record<string, unknown>): Promise<unknown>;
  reflect(req: Record<string, unknown>): Promise<unknown>;
  checkpoint(req: Record<string, unknown>): Promise<unknown>;
}

/** Memory disabled — the default. Every method is a no-op / empty. */
export class NoopHarnessMemory implements HarnessMemory {
  async recall(): Promise<string[]> {
    return [];
  }
  async recordOutcome(): Promise<void> {}
  async endSession(): Promise<void> {}
}

/** Render recalled snippets as a compact, clearly-delimited system-prompt section. */
export function formatRecallBlock(snippets: string[]): string {
  const lines = snippets.map((s) => `- ${s}`).join("\n");
  // Framed as passive background, NOT an instruction: recall is keyed by repo (not session), so a
  // fresh chat inherits these; an imperative "apply when relevant" made the model act on habits
  // (e.g. run ls) before the user asked for anything. Keep it reference-only.
  const header =
    "Background context from earlier work on this project, for reference only. Do NOT run tools or " +
    "take any action based on this memory unless the user's current request actually calls for it.";
  return `<prior_learnings source="mubit">\n${header}\n${lines}\n</prior_learnings>`;
}

function snippetOf(entry: unknown): string {
  if (typeof entry === "string") return entry.trim().slice(0, MAX_SNIPPET_CHARS);
  if (entry && typeof entry === "object") {
    for (const key of ["content", "text", "summary"]) {
      const v = (entry as Record<string, unknown>)[key];
      if (typeof v === "string" && v.trim()) return v.trim().slice(0, MAX_SNIPPET_CHARS);
    }
  }
  return "";
}

/** recall() may return an array or a wrapper ({entries|evidence|results|items: [...]}). */
function entriesOf(res: unknown): unknown[] {
  if (Array.isArray(res)) return res;
  if (res && typeof res === "object") {
    for (const key of ["entries", "evidence", "results", "items", "memories"]) {
      const v = (res as Record<string, unknown>)[key];
      if (Array.isArray(v)) return v;
    }
  }
  return [];
}

/**
 * Mubit-SDK-backed memory: session-scoped writes, project-scoped recall, fail-open.
 * `client` is any object matching MemoryClient (the real @mubit-ai/sdk Client does).
 */
export class MubitHarnessMemory implements HarnessMemory {
  constructor(
    private readonly client: MemoryClient,
    private readonly sessionId: string,
    private readonly agentId = "minima-harness",
  ) {}

  async recall(task: string, limit = MAX_SNIPPETS): Promise<string[]> {
    if (!task.trim()) return [];
    const cap = Math.max(1, Math.min(limit, MAX_SNIPPETS));
    try {
      // The SDK requires a session_id (or run_id) for recall. We use a STABLE per-repo session
      // id (see createMubitMemory), so recall surfaces prior outcomes accumulated on this repo
      // across runs — the point of a coding agent's memory — and writes land under the same id.
      const res = await this.client.recall({
        query: task,
        limit: cap,
        entry_types: RECALL_ENTRY_TYPES,
        session_id: this.sessionId,
      });
      const out: string[] = [];
      for (const entry of entriesOf(res)) {
        const text = snippetOf(entry);
        if (text) out.push(text);
        if (out.length >= cap) break;
      }
      return out;
    } catch {
      return []; // recall must never break a run
    }
  }

  async recordOutcome(record: OutcomeRecord): Promise<void> {
    if (!record.recommendationId) return;
    const q = record.quality === null ? "n/a" : record.quality.toFixed(3);
    const trace =
      `model=${record.modelId} outcome=${record.outcome} quality=${q} ` +
      `cost_usd=${record.costUsd.toFixed(6)} latency_ms=${record.latencyMs} turns=${record.turns} ` +
      `:: ${record.task.trim().slice(0, 160)}`;
    try {
      // Trace, attributed to the recommendation for provenance + dedup.
      await this.client.remember({
        content: trace,
        intent: "trace",
        session_id: this.sessionId,
        agent_id: this.agentId,
        idempotency_key: record.recommendationId,
        lane: HARNESS_LANE,
      });
      // Close the loop with a real outcome score tied to the recommendation. Never
      // fabricate — a judge abstention (quality === null) records the trace but no score.
      if (record.quality !== null) {
        await this.client.recordOutcome({
          signal: record.quality,
          outcome: record.outcome,
          reference_id: record.recommendationId,
          session_id: this.sessionId,
        });
      }
    } catch {
      // write-back must never break a run
    }
  }

  async endSession(): Promise<void> {
    try {
      await this.client.reflect({ session_id: this.sessionId });
      await this.client.checkpoint({
        context_snapshot: "",
        label: "session-end",
        session_id: this.sessionId,
      });
    } catch {
      // session teardown must never break shutdown
    }
  }
}
